display_list/display_list_color_source.h

```diff
@@ -636,6 +655,81 @@ class DlSweepGradientColorSource final : public DlGradientColorSourceBase {
   FML_DISALLOW_COPY_ASSIGN_AND_MOVE(DlSweepGradientColorSource);
 };
 
+class DlRuntimeEffectColorSource final : public DlColorSource {
+ public:
+  DlRuntimeEffectColorSource(
+      sk_sp<SkRuntimeEffect> runtime_effect,
+      std::vector<std::shared_ptr<DlColorSource>> samplers,
+      sk_sp<SkData> uniform_data)
+      : runtime_effect_(std::move(runtime_effect)),
+        samplers_(std::move(samplers)),
+        uniform_data_(std::move(uniform_data)) {}
+
+  const DlRuntimeEffectColorSource* asRuntimeEffect() const override {
+    return this;
+  }
+
+  std::shared_ptr<DlColorSource> shared() const override {
+    return std::make_shared<DlRuntimeEffectColorSource>(
+        runtime_effect_, samplers_, uniform_data_);
+  }
+
+  DlColorSourceType type() const override {
+    return DlColorSourceType::kRuntimeEffect;
+  }
+  size_t size() const override { return sizeof(*this); }
+
+  bool is_opaque() const override { return false; }
+
+  const sk_sp<SkRuntimeEffect> runtime_effect() const {
+    return runtime_effect_;
+  }
+  const std::vector<std::shared_ptr<DlColorSource>> samplers() const {
+    return samplers_;
+  }
+  const sk_sp<SkData> uniform_data() const { return uniform_data_; }
+
+  sk_sp<SkShader> skia_object() const override {
+    if (!runtime_effect_) {
+      return nullptr;
+    }
+    std::vector<sk_sp<SkShader>> sk_samplers(samplers_.size());
+    for (size_t i = 0; i < samplers_.size(); i++) {
+      sk_samplers[i] = samplers_[i]->skia_object();
+    }
+    return runtime_effect_->makeShader(uniform_data_, sk_samplers.data(),
+                                       sk_samplers.size());
+  }
+
+ protected:
+  bool equals_(DlColorSource const& other) const override {
+    FML_DCHECK(other.type() == DlColorSourceType::kRuntimeEffect);
+    auto that = static_cast<DlRuntimeEffectColorSource const*>(&other);
+    if (runtime_effect_ != that->runtime_effect_) {
+      return false;
+    }
+    if (uniform_data_ != that->uniform_data_) {
```

Comment:
Is this meant to compare their contents? Does it end up comparing their addresses instead?

Reply:
This compares if they're pointing to the same SkData - I think that's ok for now.